File: Fortran/01_Recap/01_Simple_send/wave.f90
PROGRAM wave

  USE mpi
  IMPLICIT NONE

  INTEGER, PARAMETER :: tag = 100

  INTEGER :: rank, recv_rank
  INTEGER :: nproc
  INTEGER :: left, right
  INTEGER :: ierr

  CALL MPI_Init(ierr)

  CALL MPI_Comm_size(MPI_COMM_WORLD, nproc, ierr)
  CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)

  !Set up periodic domain
  left = rank - 1
  IF (left < 0) left = nproc - 1
  right = rank + 1
  IF (right > nproc - 1) right = 0

  IF (rank == 0) THEN
    CALL MPI_Ssend(rank)
    CALL MPI_Recv(recv_rank)
  ELSE
    CALL MPI_Recv(recv_rank)
    CALL MPI_Ssend(rank)
  END IF

  PRINT *,"Rank ", rank, " got message from rank ", left, " of ", recv_rank

  CALL MPI_Finalize(ierr)

END PROGRAM wave

File: C/01_Recap/01_Simple_send/wave.c
#include <stdio.h>
#include <mpi.h>

#define TAG 100

int main(int argc, char ** argv)
{

  int rank, recv_rank, nproc, left, right;

  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &nproc);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  //Set up periodic domain
  left = rank - 1;
  if (left < 0) left = nproc - 1;
  right = rank + 1;
  if (right > nproc - 1) right = 0;

  //Complete the calls to MPI_Ssend to send rank to the processor at rank
  // "right"
  if (rank == 0) {
    MPI_Ssend(&rank);
    MPI_Recv(&recv_rank);
  } else {
    MPI_Recv(&recv_rank);
    MPI_Ssend(&rank);
  }

  printf("Rank %3d got message from rank %3d of %3d\n", rank, left, recv_rank);

  MPI_Finalize();

}
